apps/passport-server/src/services/generic-issuance/SemaphoreGroupProvider.ts

```diff
@@ -0,0 +1,309 @@
+import { SemaphoreGroupConfig } from "@pcd/passport-interface";
+import {
+  SerializedSemaphoreGroup,
+  deserializeSemaphoreGroup,
+  serializeSemaphoreGroup
+} from "@pcd/semaphore-group-pcd";
+import { uuidToBigInt } from "@pcd/util";
+import { Group } from "@semaphore-protocol/group";
+import { IPipelineConsumerDB } from "../../database/queries/pipelineConsumerDB";
+import { IPipelineSemaphoreHistoryDB } from "../../database/queries/pipelineSemaphoreHistoryDB";
+import { traced } from "../telemetryService";
+import {
+  PipelineSemaphoreGroups,
+  makeGenericIssuanceSemaphoreGroupUrl
+} from "./capabilities/SemaphoreGroupCapability";
+
+/**
+ * When a pipeline wants to trigger an update for a Semaphore group, it passes
+ * an array of these objects to the Semaphore group provider, enabling matching
+ * to group configuration and lookup of Semaphore ID from the consumer DB via
+ * the email address attached to a ticket.
+ */
+export interface SemaphoreGroupTicketInfo {
+  eventId: string;
+  productId: string;
+  email: string;
+}
+
+const SEMAPHORE_GROUP_DEPTH = 16;
+const LOG_NAME = "SemaphoreGroupProvider";
+
+/**
+ * Provides reusable Semaphore group features for pipelines.
+ */
+export class SemaphoreGroupProvider {
+  // The semaphore group configuration from the pipeline definition.
+  private groupConfigs: SemaphoreGroupConfig[];
+  private consumerDB: IPipelineConsumerDB;
+  private semaphoreHistoryDB: IPipelineSemaphoreHistoryDB;
+  private pipelineId: string;
+  // The in-memory representation of the groups, keyed by group ID.
+  private groups: Map<string, Group>;
+
+  public constructor(
+    pipelineId: string,
+    groupConfigs: SemaphoreGroupConfig[],
+    consumerDB: IPipelineConsumerDB,
+    semaphoreHistoryDB: IPipelineSemaphoreHistoryDB
+  ) {
+    this.groupConfigs = groupConfigs;
+    this.consumerDB = consumerDB;
+    this.pipelineId = pipelineId;
+    this.semaphoreHistoryDB = semaphoreHistoryDB;
+    // Groups are empty by default, but should be loaded from the history
+    // tables in the DB using `loadFromHistory()`
+    this.groups = new Map();
+  }
+
+  /**
+   * Get a group from the in-memory representation.
+   */
+  public getGroup(groupId: string): Group | undefined {
+    return this.groups.get(groupId);
+  }
+
+  /**
+   * Get a group in serialized format.
+   */
+  public async getSerializedGroup(
+    groupId: string
+  ): Promise<SerializedSemaphoreGroup | undefined> {
+    const group = await this.semaphoreHistoryDB.getLatestHistoryForGroup(
+      this.pipelineId,
+      groupId
+    );
+    if (!group) {
+      return undefined;
+    }
+    const serializedGroup = JSON.parse(group.serializedGroup);
+    return serializedGroup;
+  }
+
+  /**
+   * Get the latest root hash for a group.
+   */
+  public async getGroupRoot(groupId: string): Promise<string | undefined> {
```

Comment:
worth naming this + other functions, where appropriate, to include the word `latest`?